aiida/common/warnings.py

```diff
@@ -32,3 +34,17 @@ class AiidaTestWarning(Warning):
     """
     Class for warnings concerning the AiiDA testing infrastructure.
     """
+
+
+def warn_deprecation(message: str, version: int, stacklevel=2) -> None:
+    """Warns about a deprecation for a future aiida-core version.
+
+    Warnings are activated if the `AIIDA_WARN_v{major}` environment variable is set to `True`.
```

Comment:
Why does it work like this though? Why do we need an env variable to tell the code which deprecation warnings to print instead of just printing all of them?

Reply:
So the plugin developers and users can choose when they want to switch them on, usually when they get round to refactoring their code.
This is exactly how other libraries do their deprecations: https://docs.sqlalchemy.org/en/14/changelog/migration_20.html#migration-to-2-0-step-two-turn-on-removedin20warnings

Reply:
You "turn-on" the deprecations when running your test suite, then apply all the fixes, until there are no more warnings, job done.
This is exactly how I updated aiida-core to the sqlalchemy v2 API, and how I updated all the code in these PRs

Reply:
Basically, we don't want to be flooding users with loads of deprecation warnings for v3, as soon as they update to v2

Reply:
Mmm, I see. I agree that it is good to give them the option to silence them, but should they not be "on" by default? I would bet a lot of people may have AiiDA scripts that need update and are not necessarily being tested in this CI automated way. If we don't notify them by default they run the risk of only noticing this once the deprecated features are dropped and then they get errors instead of warnings.

On the other hand, why do we need to distinguish between versions for this (`AIIDA_WARN_v{major}`)? Why not have just `AIIDA_DEPRECATION_WARNINGS`? At any given time there will only be one set of deprecations, why do we need the selectivity?

Reply:
Because they are only just upgrading to v2, and likely won't want to be hit with lots of deprecation warnings also for v3. These can be turned on by default, in a later v2 release, closer to a v3 release.